### PR TITLE
[WIP] Annotating a few basic React and Styled components with types

### PR DESCRIPTION
**Pull request by akiselev**
This PR got a away from me, it seems ¯\_(ツ)_/¯

### Just rename `*.js` to `*.ts` or `*.jsx` to `*.tsx`

note that some JS files are actually parsed as JSX so they should be renamed to `*.tsx` going forward

---

### Inference works

Nuff said. Set `typescript.tsserver.experimental.enableProjectDiagnostics` to `true` in your VSCode config for project wide diagnostics (the `Problems` panel in VSCode will show errors in all files, instead of just the ones you have open)

---

### Most of the flow types *just work*

There are a few minor syntax differences but the transformation is largely mechanical. AFAICT (in flow) `memberName: ?FooType`, `memberName?: ?FooType`, and `memberName?: FooType` can be explained as some combination of (in Typescript this time) `memberName?: FooType` and `memberName: FooType | null`.

Function definitions require argument names, i.e. (in flow) `type FooType<T> = { funcName: (T) => void }` becomes (in Typescript):
  ```typescript
  type FooType<T> = {
    funcName: (argName: T) => void;
  }
  ```

React component Prop and State types need to be hoisted to the `React.Component<PROPS, STATE>` generic type parameters. The following (in Flow)
  ```typescript
  export default class ActionButton extends Component {
    props: Props;
    state: State;
  }
  ```

becomes (in Typescript)
  ```typescript
  export default class ActionButton extends Component<Props, State> {
     constructor(props: Props) { /* ... */ ]
  }
  ```

---

### Type Virality

When a Typescript file tries to import a Javascript file that hasn't been ported to Typescript, it throws this error:

image

That means to pass type checking, every module imported by a Typescript file must have some sort of typing. The easiest way to avoid this is by annotating from the leaves inward. See  #18326 for a naive helper in the `parse-deps.js` script, which sorts the source files based on the number of other modules they import.

*TODO:* Add notes about alternative options here i.e. topological sort

### `@types` is a wild west

Many things in `@types/*` will be out of date (*cough*React*cough*), not follow the semver of the library it's annotating, do something worldending like specify `"*"` for a common dependency, and/or eat your favorite keyboard. React's types took years to work out and since we're a major version behind we're likely even worse off on the types.  *TODO:* Explain further.

### Did I just timeout again?

Need to keep an eye on rate of timeouts of our unit test suites. Typescript has a nasty habit of breaking the camel's back.

---

### Type checking in CI/CD or continuous git history. Pick one!?!?

As you can see in the changes, the `Icon.jsx` => `Icon.tsx` and `promise.js` => `promise.ts`  change was not picked up as a rename + modification but as a deletion + creation. This breaks the continuity on git history and **must be avoided** - `git blame` is really useful for figuring out who to poke when you're trying to figure out why some code is written the way it is.

I tried to redo the commits to rename `Icon.jsx` to `Icon.tsx` in one commit and make the modifications in another commit and in my VSCode UI, it looked like a Rename+Merge. Commit `24f380f` has just the rename and `d29be80` the modification but on Github there's a full deletion and addition in the PR.

The root cause is converting a class component to a functional one, which modified whitespace for most of the file. I suspect that there are subtle differences in the implementation or configuration of Github's algorithms for detecting change threshold. (see docs for the `-M` option in `git log -M50 -p --stat`), which is why it *Worked On My Machine*®

**TLDR: It looks like we have a catch 22.** Best case scenario squashed PRs are treated differently so we can use one PR to rename from `js[x]` to `ts[x]` but any naive rename will almost certainly break typechecking. We can combine the type annotation commit with the rename commit but then we might lose git history!

Alternatively: Ban certain refactors from occuring in the same PR, i.e. can't add type annotations *and* convert a Class component to a functional one at the same time.

Stayed tuned for the next issue of "Wait, What!?"

---

## Peanut gallery

Building a mental list of common errors

### Nested type error with inherited styled components

image

Caused by `Button` inheriting from `BaseButton` styled component but `BaseButton` doesn't have an `onClick` prop defined: 
image

---

### `style` prop mismatches

See section on the wild west above.

image

---

### They said `null` was Java's original sin... so Javascript declared, "Hold my beer!"

image

### CODE REVIEW

**Review by alxnddr — COMMENTED** [review]
@akiselev looks great! Thanks for the deeper look and investigation.
I left a couple of suggestions, please take a look.
